core/trino-main/src/main/java/io/trino/operator/MultiChannelGroupByHash.java

```diff
@@ -140,23 +131,21 @@ public MultiChannelGroupByHash(
         PagesHashStrategyFactory pagesHashStrategyFactory = joinCompiler.compilePagesHashStrategyFactory(this.types, outputChannels.build());
         hashStrategy = pagesHashStrategyFactory.createPagesHashStrategy(this.channelBuilders, this.precomputedHashChannel);
 
-        startNewPage();
+        pageBuilder = new PageBuilder(types);
```

Comment:
> ince the hash table can only store
2^30 positions, the structure can be flattened to a single huge page.

Retained size of single large page will be greater than if multiple pages are used (because we use `currentPageBuilder.newPageBuilderLike()` which is adaptive)

Cost of rezising page is neglibable? What if page consists of long strings?

Reply:
> Retained size of single large page will probably be greater than if multiple pages are used

That is true. However, the size of hash table and other primitive arrays shrinks from ~37 bytes/row to ~9 bytes which gives us some headroom.
Just to play with numbers some more:
Block is resized by 1.5 so on average there is a 25% additional space retained. Wasted space for multiple-page approach is negligible for big enough number of groups.
So there is a breakeven for rows of average size (37-9) / 0.25 = 112B
So for rows lower than that we use less memory, but for bigger ones we use more. There is tradeoff here and we need to choose whether we accept it.

> Cost of rezising page is neglibable?

I guess it is not bigger than handling and allocating multiple pages, especially for multi-row case.

> What if page consists of long strings?

Aggregating by text is not necessarily the smartest thing. You can make a query fail in a much simpler way.

Reply:
> That is true. However, the size of hash table and other primitive arrays shrinks from ~37 bytes/row to ~9 bytes which gives us some headroom.

That is only valid assumption with primitive arrays

> I guess it is not bigger than handling and allocating multiple pages, especially for multi-row case.

It also depends on not having wide rows.

> Aggregating by text is not necessarily the smartest thing. You can make a query fail in a much simpler way.

There are already such aggregations in our favorite benchmark. We also insert deduplicating aggregations or aggregations for correlated subqueries synthetically which store entire rows. Customers sometimes have very wide rows (e.g. hundreds of columns).

That's why I'm asking if similar effect could be obtained by increasing page size which by default is pretty low https://github.com/trinodb/trino/pull/12072/commits/9f4dbe8e377244e073f9516522e78cbc86760f33#r858629278. It should reduce mem overhead, but also allow for 1-level addressing to fit into CPU cache

Reply:
> increasing page size

I don't quite follow.
Having a 2-level addressing with just a single page is still 2-level addressing. You still need to make 2 lookups per operation to get the data